lib/string.gi

```diff
@@ -1258,9 +1258,86 @@ InstallGlobalFunction(StringOfMemoryAmount, function(m)
     Append(s, units[shift+1]);
     return s;
 end);
-
-        
     
+InstallGlobalFunction(FormatTo, function(stream, s, args)
+    local pos, len, outstr, nextbrace, endbrace, id, argcounter;
+    argcounter := 1;
+    len := Length(s);
+    pos := 0;
+
+    if not IsOutputStream(stream) or not IsString(s) or not(IsList(args) or IsRecord(args)) then
+        ErrorNoReturn("Usage: FormatTo(<stream>, <string>, <list or record>)");
+    fi;
+
+    while pos < len do
+        nextbrace := Position(s, '{', pos);
+        endbrace := Position(s, '}', pos);
+        # Keep checking for '}}' until we find an '{'
+        while endbrace <> fail and endbrace < nextbrace do
+            if endbrace + 1 <= len and s[endbrace + 1] = '}' then
+                AppendTo(stream, s{[pos+1..endbrace]});
+                pos := endbrace + 1;
+                endbrace := Position(s, '}', pos);
+            else
+                ErrorNoReturn("Mismatched '}' at position ",endbrace);
+            fi;
+        od;
+
+        if nextbrace = fail then
+            AppendTo(stream, s{[pos+1..len]});
+            return;
+        fi;
+
+        AppendTo(stream, s{[pos+1..nextbrace-1]});
+
+        if nextbrace+1 <= len and s[nextbrace+1] = '{' then
+            AppendTo(stream, "{");
+            pos := nextbrace + 1;
+        else
+            if endbrace = fail then
+                ErrorNoReturn("Invalid format string, no matching '}' at position ", nextbrace);
+            fi;
+            id := s{[nextbrace+1..endbrace-1]};
+            if id = "" then
+                if not IsList(args) then
+                    ErrorNoReturn("data argument of Format must be a list when using {}");
+                fi;
+                if Length(args) < argcounter then
+                    ErrorNoReturn("out of bounds in Format -- asked for ",argcounter," {} when there are only ",Length(args), " arguments");
+                fi;
+                AppendTo(stream, String(args[argcounter]));
+                argcounter := argcounter + 1;
+            elif Int(id) <> fail then
+                if not IsList(args) then
+                    ErrorNoReturn("data argument of Format must be a list when using {<integer>}");
+                fi;
+                if Int(id) < 1 or Int(id) > Length(args) then
+                    ErrorNoReturn("out of bounds in Format -- asked for {",Int(id),"} when there are only ",Length(args), " arguments");
+                fi;
+                AppendTo(stream, String(args[Int(id)]));
```

Comment:
`String` is problematic because not many objects support `String` and just return `<object>`, also one could argue that `String` builds up a buffer, whereas `PrintTo` or `AppendTo` let the object's implementor decide what to do.

why not just `args[Int(id)]`? 

There is of course the question whether you want to support `String`, `Display` and `View` as format strings...

Reply:
See also https://github.com/gap-system/gap/issues/1829

Reply:
I'm happy to pick which we want (string, display or view), but if you just pass in strings then of course they'll be printeed fine (I could check `IsString` to avoid the extra call), but it seems a pain to make people call `String` if they just want to pass in say a permutation.

Reply:
I'm not sure about DisplayString (they generally don't compose very well) but one way or another it could be useful to be able to include either the ViewString or the String of an object, depending on what you want the formatted string for. Could be an option encoded somehow in the format-string, or a second function FormatView (or some better name).

Reply:
For consistency with names elsewhere the function should ideally be a noun describing the result, rather than a verb which suggests it modifies its argument. We have `FormattedString` but it's marked obsolete, so could possibly be resused.

Reply:
Re String/Display/View: In a `printf` style function, I'd handle that with different placeholders in the format string, say:  `Format("%{display} vs. %{view} vs. %{string}", myObj, myObj, myObj);` to get a comparison of the three strings.

Looking at https://pyformat.info, I guess the "pythonic" way would be to use a "value conversion" suffix, like  `Format("{0!d} vs. {0!v}. vs {0!s}", myObj)`. If we default to using `String`, then there could also be something to force "raw" output, i.e. no conversion. Or conversely, we could default to *not* using `String` (as @markuspf suggested), but still allow people to enable it via a value conversion suffix.